src/unix/netmap-udp.c

```diff
@@ -0,0 +1,735 @@
+
+    if (req->bufs != req->bufsml)
+      uv__free(req->bufs);
+    req->bufs = NULL;
+
+    if (req->send_cb == NULL)
+      continue;
+
+    if (req->status >= 0)
+      req->send_cb(req, 0);
+    else
+      req->send_cb(req, req->status);
+  }
+
+  if (QUEUE_EMPTY(&loop->netmap->write_queue) && !removed) {
+    uv__io_stop(loop, &loop->netmap->io_watcher, POLLOUT);
+    if (!uv__io_active(&loop->netmap->io_watcher, POLLIN)) {
+      uv__handle_stop(loop->netmap);
+    }
+  }
+
+  loop->netmap->flags &= ~UV_HANDLE_UDP_PROCESSING;
+}
+
+static void uv__udp_netmap_io(uv_loop_t* loop, uv__io_t* w, unsigned int revents) {
+  int i;
+  uint64_t j, len;
+  struct netmap_ring* ring;
+
+  if (loop->netmap == NULL) {
+    return;
+  }
+  if (revents & POLLIN) {
+    for (i = loop->netmap->intf->first_rx_ring; i <= loop->netmap->intf->last_rx_ring; i++) {
+      ring = NETMAP_RXRING(loop->netmap->intf->nifp, i);
+
+      len = nm_ring_space(ring);
+      // todo consider limiting len here so that we can run timers etc
+      for (j = 0; j < len; j++) {
+        struct netmap_slot* slot;
+        uint8_t* p;
+
+        slot = &ring->slot[ring->cur];
+        p = NETMAP_BUF(ring, slot->buf_idx);
+
+        uv__udp_netmap_recv_packet(loop, slot, p);
+
+        ring->head = ring->cur = nm_ring_next(ring, ring->cur);
+      }
+    }
+  }
+
+  if ((revents & POLLOUT)) {
+    // only tx on our rx ring, and only tx for as long as there's room to do so
+    for (i = loop->netmap->intf->first_rx_ring; i <= loop->netmap->intf->last_rx_ring; i++) {
+      uint64_t pkt_num;
+
+      ring = NETMAP_TXRING(loop->netmap->intf->nifp, i);
+      len = nm_ring_space(ring);
+      pkt_num = 0;
+      while (1) {
+        QUEUE* q;
+        uv_udp_send_t* req;
+
+        if (pkt_num == len) {
+          break;
+        }
+
+        if (QUEUE_EMPTY(&loop->netmap->write_queue)) {
+          break;
+        }
+
+        q = QUEUE_HEAD(&loop->netmap->write_queue);
+        req = QUEUE_DATA(q, uv_udp_send_t, queue);
+
+        uv__udp_netmap_send_udp(loop, req, ring);
+        uv__io_feed(loop, &loop->netmap->io_watcher);
+
+        pkt_num++;
+      }
+    }
+    uv__udp_netmap_run_completed(loop);
+  }
+}
+
+int uv_udp_netmap_init(uv_loop_t* loop, const char *fname) {
+  if (loop->netmap != NULL) {
+    return -1;
+  }
+
+  loop->netmap = uv__malloc(sizeof(uv_netmap_t));
+  memset(loop->netmap, 0, sizeof(uv_netmap_t));
+
+  nm_desc_t* netmap_desc = nm_open(fname, NULL, 0, 0);
+  if (netmap_desc == NULL) {
+    printf("netmap error\n");
+    return -1;
+  }
+
+  loop->netmap->intf = netmap_desc;
+
+  uv__handle_init(loop, (uv_handle_t*)loop->netmap, UV_NETMAP);
+  uv__io_init(&loop->netmap->io_watcher, uv__udp_netmap_io, loop->netmap->intf->fd);
+  QUEUE_INIT(&loop->netmap->write_queue);
+  QUEUE_INIT(&loop->netmap->write_completed_queue);
+
+  return 0;
+}
+
+static uv_udp_netmap_close_cb(uv_handle_t* handle)
+{
+  handle->loop->netmap = NULL;
+  free(handle);
```

Comment:
`uv__free`? This is freeing the netmap struct, yes? The line above is effectively `netmap(handle)->loop->netmap`?